### ISSUE TITLE
Proposal for reworking PreparedStatement

### ISSUE THREAD

**Issue by BillyYccc** [eclipse-vertx/vertx-sql-client#687]
## Context

Some time ago I made a change that `PreparedStatement` is not cached on the client any more(see https://github.com/eclipse-vertx/vertx-sql-client/commit/e53235cbe2bc2fd96aabe35350ab05ba840b164f), the motivation of that change is to let users control the lifecycles of `PreparedStatement` and avoid lifecycle conflicts with one-shot preparedQueries, the user-created `PreparedStatement` might share a same statement handle with the handle used by the one-shot preparedQueries internally, for example when the prepared statement is evicted from the cache and automatically closed by the client, it will unexpectedly close the relative `PreparedStatement` at the same time. Therefore I chose to have users keep a reference to the object, cache the `PreparedStatement` object themselves and it's not cached by the client any more but now I find the design very limited. 

A biggest problem of this design is that the statement is escaped from the scope of the connection it belongs to. Likely said once a connection is disconnected the `PreparedStatement` is not usable any more and there is no way for users who cache the `PreparedStatement` object to know that. Additionally it's nearly impossible to chain executions of cached `PreparedStatement` and other behaviors in the same connection/session because users never know which connection the `PreparedStatement` is running on. 

## Proposal

In this proposal in order to address those issues above I would like to bring back the `PreparedStatement` into the client cache, so that when users get a cached/non-cached `PreparedStatement` it's guaranteed to be always accessed from a `SqlConnection`.

A usage case is like this:
```
  @Test
  public void example(SqlConnectOptions connectOptions) {
    connectOptions.setCachePreparedStatements(true);
    Pool pool = Pool.pool(connectOptions);
    pool.getConnection()
      .onSuccess(conn -> {
        conn.prepare("SELECT * FROM USERS")
          .onSuccess(stmt -> {
            // no matter whether the statement is cached or not it's always accessed from a connection
            // and the stmt is cached in the client again
            stmt.query()
              .execute()
              .onSuccess(res -> {
                conn.query("INSERT INTO tmp VALUES (1)")
                  .execute()
                  .onSuccess(res2 -> {

                  });
              });
          });
      });
  }
```

I would like to separate the handles of `PreparedStatement` and one-shot prepared-queries when caching is enabled as well. So when statement cache is enabled, the `PreparedStatement` will be cached in another cache which is different from the cache one-shot prepared-queries is using. One thing I am not certain is whether we need complicated caching strategy(like LRU or some others...) for this kind of caching usage, I would like to only provide a unlimited-size `HashMap` based implementation and when users calling `PreparedStatement#close` the statement is closed and removed from the cache.

**Comment by BillyYccc** [comment]
@vietj @aguibert I would like to know how you think about this.

**Comment by vietj** [comment]
I think your use case is valid.

However one issue with explicitly created prepared statement caching is the shared cache itself: you cannot evict a prepared statement from the shared cache when it is being used in an operation that spans several interactions like a cursor.

One-shot prepared statements lends well to caching because they do not provide cursor operations and we can easily uncache them as needed. Actually this cache is important for one-shot operations because otherwise it would be less efficient as the user has no way to keep an handle on the prepared query statement.

**Comment by BillyYccc** [comment]
yes the cache should be different from what one-shot preparedQueries are using. I would like to make the things simple so the cache should be a `HashMap` and there won't be a size limit and any automatically eviction here for the cache, intention for enabling the cache is only to let users able to reuse the handle.

**Comment by vietj** [comment]
do we have examples of JDBC client pool behaving this way ? i.e caching prepared statement between connection use ?

**Comment by vietj** [comment]
if we go this way, then I think this would be more something belonging to the pooled connection part of the code base and not the actual socket connection that should not care about this concern, in addition this concern is mostly valid for pooled connections, i.e if you connect directly then it does not make sense.

**Comment by BillyYccc** [comment]
I don't understand what're the relations between connection statement cache and pooled connections.

For JDBC this is an example of chain operation usage(i.e. a simple query + cache prepare)
```
    private void example(Connection conn) throws Exception {
        Statement simpleQuery = conn.createStatement();
        ResultSet resultSet0 = simpleQuery.executeQuery("SELECT 1");
        System.out.println(resultSet0.next());
        System.out.println(resultSet0.getInt(1));
        simpleQuery.close();
        String sql = "SELECT * FROM datatype WHERE id = ?";
        PreparedStatement preparedStatement = conn.prepareStatement(sql); // if cache option is enabled, the `PreparedStatement` is retrieved from the cache
        preparedStatement.setString(1,"1");
        ResultSet resultSet = preparedStatement.executeQuery();
        System.out.println(resultSet.next());
        System.out.println(resultSet.getString("Text"));
        preparedStatement.close(); // close will do nothing but put the statement back to the cache
    }
```

**Comment by vietj** [comment]
as I understand the goal is that cached prepared statement survive between two connection pool usage of the same physical connection so they don't need to be prepared again.
